### PR TITLE
[routing] Show in navigation destination tags

### PR DESCRIPTION
**Pull request by AntonM030481**
1st iteration for #849

Previously only name of next street was shown.

Now 
- it will show "[ref] name" for next street.
- for highway links it will show "[target:ref] target".

As before we go forward through segments to find fist named one (but this is limited to 400m).

Test output can be seen in Debug output of qt::DrawWidget::SubmitFakeLocationPoint().
E.g. at (34.68730, 32.91658) -> (34.69269, 32.92118)

Suggested next step is to replace [ref] and [target:ref] by icons generated by GetRoadShields().

### CODE REVIEW

File: routing/directions_engine.cpp [hunk 1]
```diff
@@ -66,7 +66,16 @@ void DirectionsEngine::LoadPathAttributes(FeatureID const & featureId,
   pathSegment.m_onRoundabout = ftypes::IsRoundAboutChecker::Instance()(*ft);
   pathSegment.m_isOneWay = ftypes::IsOneWayChecker::Instance()(*ft);
 
-  pathSegment.m_name = ft->GetName(StringUtf8Multilang::kDefaultCode);
+  string name;
+  if (!ft->GetRoadNumber().empty())
+    name = "[" + ft->GetRoadNumber() + "] ";
```
Comment (vng) [inline routing/directions_engine.cpp:71]: Use var:
```
if (auto const & rn = ft->GetRoadNumber(); !rn.empty())
  name += rn;
```

File: routing/directions_engine.cpp [hunk 2]
```diff
@@ -66,7 +66,16 @@ void DirectionsEngine::LoadPathAttributes(FeatureID const & featureId,
...
+  name += ft->GetName(StringUtf8Multilang::kDefaultCode);
+  name = ft->GetName(StringUtf8Multilang::kDefaultCode);
```
Comment (vng) [inline routing/directions_engine.cpp:73]: ? Leave one statement.

File: routing/directions_engine.cpp [hunk 3]
```diff
@@ -66,7 +66,16 @@ void DirectionsEngine::LoadPathAttributes(FeatureID const & featureId,
...
+  if (name.empty())
+    if (!ft->GetMetadata(feature::Metadata::FMD_DESTINATION_REF).empty())
+    name = "[" + string(ft->GetMetadata(feature::Metadata::FMD_DESTINATION_REF)) + "] ";
+    name += ft->GetMetadata(feature::Metadata::FMD_DESTINATION);
```
Comment (vng) [inline routing/directions_engine.cpp:77]: This looks like strange python code :) Use var and block.

File: routing/directions_engine.cpp [hunk 4]
```diff
@@ -66,7 +66,16 @@ void DirectionsEngine::LoadPathAttributes(FeatureID const & featureId,
...
+  pathSegment.m_name = name;
```
Comment (vng) [inline routing/directions_engine.cpp:78]: ... = std::move(name)

**Review by vng — CHANGES_REQUESTED** [review]

File: generator/pygen/pygen.cpp [hunk 5]
```diff
@@ -347,6 +347,8 @@ BOOST_PYTHON_MODULE(pygen)
         .value("contact_twitter", Metadata::EType::FMD_CONTACT_TWITTER)
         .value("contact_vk", Metadata::EType::FMD_CONTACT_VK)
         .value("contact_line", Metadata::EType::FMD_CONTACT_LINE);
+        .value("destination", Metadata::EType::FMD_DESTINATION);
+        .value("destination_ref", Metadata::EType::FMD_DESTINATION_REF);
```
Comment (vng) [inline generator/pygen/pygen.cpp:351]: Hm, Did you get what for this pygen.cpp module?
And why "destination_ref", if we should get "destination:ref"

File: routing/directions_engine.cpp (hunk 1 above)
Comment (AntonM030481) [reply routing/directions_engine.cpp:71]: OK

File: routing/directions_engine.cpp (hunk 2 above)
Comment (AntonM030481) [reply routing/directions_engine.cpp:73]: OK. Thanks.

File: routing/directions_engine.cpp (hunk 3 above)
Comment (AntonM030481) [reply routing/directions_engine.cpp:77]: Too much Python yesterday :)

File: generator/pygen/pygen.cpp (hunk 5 above)
Comment (AntonM030481) [reply generator/pygen/pygen.cpp:351]: No, I didn't :)
I applied the same logic as for FMD_TURN_LANES_FORWARD, FMD_BUILDING_LEVELS, etc.

BTW, can't we use duplication of matching in indexer::ToString, indexer::Metadata::TypeFromString and in Pygen.
We need 2 direct and 1 reverse transformation.
E.g. create static array and initialize it as `a[Metadata::EType] = value;`
And in cycle populate `map<string, Metadata::EType>` and extend it manually where needed for reverse transformation?

File: routing/directions_engine.cpp (hunk 4 above)
Comment (AntonM030481) [reply routing/directions_engine.cpp:78]: I decided to get rid of |name|.

**Comment by AntonM030481** [comment]
@vng How can I add additional text field to qt window to show in desktop navigation mode the result of 
Debug output of qt::DrawWidget::SubmitFakeLocationPoint()?
I hope it will be useful in the future too, e.g. if we will add lane guidance.
In the perfect case it should be the copy of GUI guidance from mobile, but text version will be OK (ascii art!)

File: generator/pygen/pygen.cpp (hunk 5 above)
Comment (AntonM030481) [reply generator/pygen/pygen.cpp]: OK

**Comment by AntonM030481** [comment]
Now everything is OK.

But it will be better to make algorithm more flexible to handle incomplete data.

Currently it is expected that for link both destination and destination:ref tags will exist together.
But sometimes only destination tag exists for link. And destination:ref can be calculated by checking next segments of route until link will end and the main road will start (and sometimes it will have ref tag, missed in link).

Also it's better to go trough longer distance (compared to current 400m) looking for existing data,
if we go along the link. Sometimes links are really long (e.g. 1km+ in USA) and they have no tag it's better to go until link will end (no matter how long it is) until main road will start.

**Comment by biodranik** [comment]
These are very good ideas. Let's implement them!

**Review by vng — APPROVED** [review]

**Comment by vng** [comment]
Still in draft? Can merge it now to get "destinations" in next data snapshot.

**Comment by AntonM030481** [comment]
Let's merge now.
Additional improvements https://github.com/organicmaps/organicmaps/pull/2632#issuecomment-1140297558 are no trivial since currently we don't have info about "link" status of roads at level of Route class and straightforward adding it will be non elegant at all.

**Pull request merged**